### ISSUE TITLE
Unexpected `[no-redef]` errors for plugin-generated class replacement attributes if `[name-defined]` occurs in the same module

### ISSUE THREAD

**Issue by bzoracler** [python/mypy#17464]
**Bug Report & To Reproduce**

I have a class called `Addee` with an attribute (currently a method called `var`), which I want to replace with a plain variable (essentially: `var: int`). The buggy module and plugin implementation module is given below:

```python
# addee.py
class _Addee: ...

class Addee(_Addee):
    def var(self, /) -> None: ...

# a: asdfjkl  # Uncomment to reproduce the bug in this report
```

```python
# mypy_plugin.py
from __future__ import annotations

import typing_extensions as t

import mypy.plugin
import mypy.plugins.common

if t.TYPE_CHECKING:
    import collections.abc as cx

def plugin(version: str) -> type[mypy.plugin.Plugin]:
    return AttributeAddingPlugin

class AttributeAddingPlugin(mypy.plugin.Plugin):
    @t.override
    def get_base_class_hook(
        self, fullname: str
    ) -> cx.Callable[[mypy.plugin.ClassDefContext], None] | None:
        if fullname == "addee._Addee":
            return base_class_hook
        return None

def base_class_hook(ctx: mypy.plugin.ClassDefContext, /) -> None:
    mypy.plugins.common.add_attribute_to_class(
        ctx.api, ctx.cls, "var", ctx.api.named_type("builtins.int")
    )
```

**Expected Behavior**

```python
# addee.py
class _Addee: ...

class Addee(_Addee):
    def var(self, /) -> None: ...

a: asdfjkl  # mypy: Name "asdfjkl" is not defined  [name-defined]
```

**Actual Behaviour**

An unexpected `[no-redef]` error additionally appears; `type: ignore[name-defined]` on `a: asdfjkl` does not suppress this. Comment out `a: asdfjkl` and the error disappears.

```python
# addee.py
class _Addee: ...

class Addee(_Addee):
    def var(self, /) -> None: ...  # mypy: Name "var" already defined (possibly by an import)  [no-redef]

a: asdfjkl  # mypy: Name "asdfjkl" is not defined  [name-defined]
```

**Your Environment**

- Mypy version used: 1.10.1
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None (apart from `plugins = mypy_plugin.py`)
- Python version used: 3.10
